File: src/rate-limiter/lua/sliding-window.lua
-- Sliding Window Rate Limiter Lua Script
--
-- KEYS[1] = current window key:  rl:{tenant_id}:{current_window_timestamp}
-- KEYS[2] = previous window key: rl:{tenant_id}:{prev_window_timestamp}
--
-- ARGV[1] = limit (max requests allowed in window)
-- ARGV[2] = window size in seconds (e.g. 60)
-- ARGV[3] = current timestamp (seconds with fraction, e.g. 1727050000.123)
-- ARGV[4] = window start timestamp (seconds, e.g. 1727049960)
--
-- Returns table: { allowed (1 or 0), remaining_tokens, reset_after_seconds }

local current_key = KEYS[1]
local prev_key = KEYS[2]

local limit = tonumber(ARGV[1])
local window_size = tonumber(ARGV[2])
local now = tonumber(ARGV[3])
local current_window_start = tonumber(ARGV[4])

-- Get counts for current and previous window
local current_count = tonumber(redis.call('get', current_key) or '0')
local prev_count = tonumber(redis.call('get', prev_key) or '0')

-- Calculate position within current window: fraction from 0.0 to 1.0
local time_into_current = now - current_window_start
local weight = (window_size - time_into_current) / window_size
if weight < 0 then
  weight = 0
elseif weight > 1 then
  weight = 1
end

-- Estimated requests in sliding window = previous_window * weight + current_window
local estimated_count = math.floor(prev_count * weight + current_count)

if estimated_count < limit then
  -- Allowed: increment current window counter and set TTL (2x window size for safety)
  redis.call('incr', current_key)
  redis.call('expire', current_key, window_size * 2)

  local remaining = limit - (estimated_count + 1)
  if remaining < 0 then
    remaining = 0
  end

  local reset_after = math.ceil(window_size - time_into_current)
  return { 1, remaining, reset_after }
else
  -- Rejected: rate limit exceeded
  local reset_after = math.ceil(window_size - time_into_current)
  return { 0, 0, reset_after }
end

File: src/rate-limiter/sliding-window.ts
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import type { Redis } from 'ioredis';
import { getRedisClient } from '../redis/client.js';
import { config } from '../config/index.js';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const LUA_SCRIPT_PATH = path.join(__dirname, 'lua', 'sliding-window.lua');

export interface RateLimitResult {
  allowed: boolean;
  limit: number;
  remaining: number;
  resetAfterSeconds: number;
}

export class SlidingWindowRateLimiter {
  private redis: Redis;
  private luaScript: string;
  private scriptSha: string | null = null;
  private windowSizeSeconds: number;

  constructor(redis?: Redis, windowSizeSeconds = 60) {
    this.redis = redis ?? getRedisClient();
    this.windowSizeSeconds = windowSizeSeconds;
    this.luaScript = fs.readFileSync(LUA_SCRIPT_PATH, 'utf-8');
  }

  /**
   * Pre-load the Lua script SHA into Redis to maximize throughput.
   */
  async init(): Promise<void> {
    try {
      this.scriptSha = await this.redis.script('LOAD', this.luaScript) as string;
    } catch (err) {
      console.warn('[RateLimiter] Failed to preload Lua script SHA:', (err as Error).message);
    }
  }

  /**
   * Determine rate limit for a specific tenant tier.
   */
  getLimitForTier(tier: 'free' | 'pro' | 'enterprise'): number {
    switch (tier) {
      case 'pro':
        return config.rateLimits.pro;
      case 'enterprise':
        return config.rateLimits.enterprise;
      case 'free':
      default:
        return config.rateLimits.free;
    }
  }

  /**
   * Check and consume rate limit token for a tenant.
   *
   * @param tenantId The unique tenant identifier
   * @param tier Tenant subscription tier
   * @returns RateLimitResult with status and headers info
   */
  async consume(tenantId: string, tier: 'free' | 'pro' | 'enterprise'): Promise<RateLimitResult> {
    const limit = this.getLimitForTier(tier);
    const now = Date.now() / 1000;
    const windowStart = Math.floor(now / this.windowSizeSeconds) * this.windowSizeSeconds;
    const prevWindowStart = windowStart - this.windowSizeSeconds;

    const currentKey = `rl:${tenantId}:${windowStart}`;
    const prevKey = `rl:${tenantId}:${prevWindowStart}`;

    try {
      let result: [number, number, number];

      if (this.scriptSha) {
        try {
          result = (await this.redis.evalsha(
            this.scriptSha,
            2,
            currentKey,
            prevKey,
            limit.toString(),
            this.windowSizeSeconds.toString(),
            now.toString(),
            windowStart.toString(),
          )) as [number, number, number];
        } catch (evalShaError: unknown) {
          // If NOSCRIPT, fallback to EVAL and reload SHA
          if ((evalShaError as Error).message?.includes('NOSCRIPT')) {
            this.scriptSha = await this.redis.script('LOAD', this.luaScript) as string;
            result = (await this.redis.eval(
              this.luaScript,
              2,
              currentKey,
              prevKey,
              limit.toString(),
              this.windowSizeSeconds.toString(),
              now.toString(),
              windowStart.toString(),
            )) as [number, number, number];
          } else {
            throw evalShaError;
          }
        }
      } else {
        result = (await this.redis.eval(
          this.luaScript,
          2,
          currentKey,
          prevKey,
          limit.toString(),
          this.windowSizeSeconds.toString(),
          now.toString(),
          windowStart.toString(),
        )) as [number, number, number];
      }

      const [allowedNum, remaining, resetAfterSeconds] = result;

      return {
        allowed: allowedNum === 1,
        limit,
        remaining,
        resetAfterSeconds,
      };
    } catch (redisError) {
      // Failure mode: Fail-safe (reject request or notify outage).
      // We log and re-throw, allowing middleware to return 503 Service Unavailable.
      throw new Error(`Rate limiter storage unavailable: ${(redisError as Error).message}`);
    }
  }

  /**
   * Reset rate limit state for a tenant (useful for testing or manual admin reset).
   */
  async resetTenant(tenantId: string): Promise<void> {
    const pattern = `rl:${tenantId}:*`;
    const stream = this.redis.scanStream({ match: pattern, count: 100 });
    const keys: string[] = [];

    for await (const batch of stream) {
      if (batch.length > 0) {
        keys.push(...batch);
      }
    }

    if (keys.length > 0) {
      await this.redis.del(...keys);
    }
  }
}
